utils/fork-tree/src/lib.rs

```diff
@@ -345,18 +298,89 @@ where
 		F: Fn(&H, &H) -> Result<bool, E>,
 		P: Fn(&V) -> bool,
 	{
-		// search for node starting from all roots
-		for (index, root) in self.roots.iter().enumerate() {
-			let node = root.find_node_index_where(hash, number, is_descendent_of, predicate)?;
-
-			// found the node, early exit
-			if let FindOutcome::Found(mut node) = node {
-				node.push(index);
-				return Ok(Some(node))
+		let mut path = vec![];
+		let mut children = &self.roots;
+		let mut i = 0;
+
+		while i < children.len() {
+			let node = &children[i];
+			if node.number < *number && is_descendent_of(&node.hash, hash)? {
+				if !predicate(&node.data) {
+					break
+				}
+				path.push(i);
+				i = 0;
+				children = &node.children;
+			} else {
+				i += 1;
 			}
 		}
 
-		Ok(None)
+		Ok(if path.is_empty() { None } else { Some(path) })
+	}
+
+	/// Prune the tree, removing all non-canonical nodes. We find the node in the
+	/// tree that is the deepest ancestor of the given hash and that passes the
+	/// given predicate. If such a node exists, we re-root the tree to this
+	/// node. Otherwise the tree remains unchanged. The given function
+	/// `is_descendent_of` should return `true` if the second hash (target) is a
+	/// descendent of the first hash (base).
+	///
+	/// Returns all pruned node data.
+	pub fn prune<F, E, P>(
```

Comment:
From what I can see the behavior of this new implementation is different than the original one (unless I'm looking at it wrong and my testcase is wrong); here's a reproduction:

```rust
	#[test]
	fn test_prune() {
		fn is_descendent_of(parent: &char, child: &char) -> Result<bool, std::convert::Infallible> {
			match *parent {
				'A' => Ok(['B', 'C', 'D'].contains(child)),
				'B' => Ok(['D'].contains(child)),
				'C' => Ok(false),
				'D' => Ok(false),
				_ => unreachable!(),
			}
		}

		// A -> B -> D
		//   \> C

		let mut tree: ForkTree<char, u8, bool> = ForkTree::new();
		tree.import('A', 1, true, &is_descendent_of).unwrap();
		tree.import('B', 2, false, &is_descendent_of).unwrap();
		tree.import('C', 2, true, &is_descendent_of).unwrap();
		tree.import('D', 3, false, &is_descendent_of).unwrap();
		let nodes = tree
			.prune(&'D', &3, &is_descendent_of, &|&value| value == false)
			.map(|i| i.collect::<Vec<_>>())
			.unwrap();
		assert_eq!(nodes, vec![('A', 1, true,), ('C', 2, true,),]);
	}
```

This passes on the old implementation and fails on this new one.

Reply:
I also insert your test within the others